utils.py

```diff
@@ -0,0 +1,80 @@
+from models import db, Stats, Player
+
+class StatsCalculator:
+    def __init__(self, game_id, points):
+        self.game_id = game_id
+        self.points = points
+        self.stats = {}
+
+    def run(self):
+        for point in self.points:
+            self.process_point(point)
+
+        return self.stats.items()
+
+    def process_point(self, point):
+        events = point['events']
+        offensePlayers = point['offensePlayers']
+        defensePlayers = point['defensePlayers']
+
+        for idx, event in enumerate(events):
+            if event['type'] == 'PASS':
+                self.add_stat(event['firstActor'], 'completions')
+                self.add_stat(event['secondActor'], 'catches')
+
+            elif event['type'] == 'DROP':
+                previous_event = events[idx-1]
+                self.add_stat(event['firstActor'], 'drops')
+                self.add_stat(previous_event['secondActor'], 'threw_drops')
+
+            elif event['type'] == 'THROWAWAY':
+                self.add_stat(event['firstActor'], 'throw_aways')
+
+            elif event['type'] == 'DEFENSE':
+                self.add_stat(event['firstActor'], 'd_blocks')
+
+            elif event['type'] == 'POINT':
+                self.add_stat(event['firstActor'], 'goals')
+
+                # Assist and 2nd Assist
+                previous_previous_event = events[idx-2]
+                previous_event = events[idx-1]
+
+                if previous_event['type'] == 'PASS':
+                    self.add_stat(previous_event['firstActor'], 'assists')
+
+                    if previous_previous_event['type'] == 'PASS':
+                        self.add_stat(previous_previous_event['firstActor'], 'second_assists')
+
+                # Finish Point
+                offenseScored = event['firstActor'] in offensePlayers
+
+                if offenseScored:
+                    [self.add_stat(player, 'o_points_for') for player in offensePlayers]
+                    [self.add_stat(player, 'd_points_against') for player in defensePlayers]
+                else:
+                    [self.add_stat(player, 'o_points_against') for player in offensePlayers]
+                    [self.add_stat(player, 'd_points_for') for player in defensePlayers]
+
+            elif event['type'] == 'PULL':
+                self.add_stat(event['firstActor'], 'pulls')
+
+
+    def add_stat(self, player_name, stat):
+        player = self.get_or_create_player(player_name)
+
+        if player.name not in self.stats:
+            self.stats[player.name] = Stats(self.game_id, player.id)
+
+        self.stats[player.name].count_stat(stat)
```

Comment:
@wingleungchan so I did end up delegating the count_stat to the Stat model instance like you were suggesting. I kept all the unsaved Stat models in a hash keyed by player_name for easy access while looping through the events and then delegate the count_stat.